packages/components/rating/src/Rating.tsx

```diff
@@ -1 +1,182 @@
-export const Rating = () => <>rating</>
+import { cx } from 'class-variance-authority'
+import {
+  type ChangeEvent,
+  ComponentPropsWithoutRef,
+  forwardRef,
+  type MouseEvent,
+  PropsWithChildren,
+  useCallback,
+  useRef,
+  useState,
+} from 'react'
+
+import { RatingStar } from './RatingStar'
+import { type Size } from './types'
+import { getNearestDecimal, getStarValue, splitAt } from './utils'
+
+export interface RatingProps extends PropsWithChildren<ComponentPropsWithoutRef<'div'>> {
+  /**
+   * Use the `defaultValue` prop to set the default value of the input, on a from 0 to 5.
+   *
+   * Use this when you want to use it in an uncontrolled manner
+   */
+  defaultValue?: number
+  /**
+   * The value is the number of the rating selected, on a scale from 0 to 5.
+   *
+   * Use this when you want to use it in a controlled manner,
+   * in conjunction with the `onValueChange` prop
+   */
+  value?: number
+  /**
+   * Event handler called when the value changes.
+   */
+  onValueChange?: (value: number) => void
+  /**
+   * Sets the component as interactive or not.
+   * @default undefined
+   */
+  readOnly?: boolean
+  /**
+   * When `true`, prevents the user from interacting.
+   * @default false
+   */
+  disabled?: boolean
+  /**
+   * Sets the size of the stars.
+   * @default 'md'
+   */
+  size?: Size
+  /**
+   * Name of the underlying input.
+   * @default undefined
+   */
+  name?: string
+  /**
+   * id of the underlying input.
+   * @default undefined
+   */
+  id?: string
+  /**
+   * aria-label of the underlying input.
+   * @default undefined
+   */
+  'aria-label'?: string
+}
+
+export const Rating = forwardRef<HTMLDivElement, RatingProps>(
+  (
+    {
+      defaultValue,
+      value: propValue,
+      onValueChange,
+      size = 'md',
+      disabled,
+      readOnly,
+      name,
+      id,
+      ...rest
+    },
+    ref
+  ) => {
+    const [ratingValue, setRatingValue] = useState(defaultValue ?? propValue)
+    const inputRef = useRef<HTMLInputElement>(null)
+    const starRefList = useRef<HTMLDivElement[]>([])
+
+    const value = propValue ?? ratingValue
+
+    const isControlled = defaultValue === undefined
+    const isNonInteractive = !!(disabled || readOnly)
+
+    function onStarClick(index: number) {
+      if (!inputRef.current || isNonInteractive) return
+
+      if (isControlled) onValueChange?.(index + 1)
+      else setRatingValue(index + 1)
+
+      inputRef.current.focus()
+      inputRef.current.setAttribute('data-clicked', '')
+    }
+
+    function onInputChange(event: ChangeEvent<HTMLInputElement>) {
+      if (isNonInteractive) return
```

Comment:
To simplify each function, wdyt of moving this logic such as:
```tsx
  {...!isNonInteractive && ({ onChange: onInputChange })}
```
? :thinking:

Reply:
Yes, I had that in mind initially. It's just that this syntax is ugly in my opinion. However, deep down, I find it more explicit to not call a function when a predicate is not met, rather than calling the function every time and checking inside

Will update it 👍